### PR TITLE
*: Add more CPU profiler metrics

### PR DESCRIPTION
**Pull request by kakkoyun**
- [Add metrics to track BPF events](https://github.com/parca-dev/parca-agent/pull/2041/commits/4b13c37fcec91b88e666dec330cfa681f32ca01b)
- [Refactor BPF metric collect package](https://github.com/parca-dev/parca-agent/pull/2041/commits/61825d5abe72a70d7fb30b230a8be20fe266bcdb)
- [Reorgranize BPF packages](https://github.com/parca-dev/parca-agent/pull/2041/commits/ddede941cbbcf8abc4e9b8b63462d87c934237d7)
- [Move everything under bpf/](https://github.com/parca-dev/parca-agent/pull/2041/commits/49ff1ff0497192739995836eeb8ef60667b11fc0)
- [Add more metrics to track unwind table operations](https://github.com/parca-dev/parca-agent/pull/2041/commits/702fefa2a191685ebf24532673177d7831da1cdf)

### Why?

We want to track more metrics for the failure modes of the BPF code.

This PR also includes some code re-organization to separate components semantically.

### What?

Refactor the CPU profiler and the BPF-related code to use new packages and improve logging, error handling, and metrics. 

### Test Plan

- `make test/profiler`
- CI

<details><summary>Relevant metrics</summary>
<p>

```

# HELP parca_agent_native_unwinder_error_total There was an error while unwinding the stack.
# TYPE parca_agent_native_unwinder_error_total counter
parca_agent_native_unwinder_error_total{reason="catchall"} 0
parca_agent_native_unwinder_error_total{reason="frame_pointer_action"} 1
parca_agent_native_unwinder_error_total{reason="jit_mixed_mode_disabled"} 0
parca_agent_native_unwinder_error_total{reason="jit_unupdated_mapping"} 0
parca_agent_native_unwinder_error_total{reason="pc_not_covered"} 17
parca_agent_native_unwinder_error_total{reason="pc_not_covered_jit"} 0
parca_agent_native_unwinder_error_total{reason="should_never_happen"} 0
parca_agent_native_unwinder_error_total{reason="truncated"} 0
parca_agent_native_unwinder_error_total{reason="unsupported_cfa_register"} 0
parca_agent_native_unwinder_error_total{reason="unsupported_expression"} 0
# HELP parca_agent_native_unwinder_samples_total Total samples.
# TYPE parca_agent_native_unwinder_samples_total counter
parca_agent_native_unwinder_samples_total{unwinder="dwarf"} 42
# HELP parca_agent_native_unwinder_success_total Samples that unwound successfully reaching the bottom frame.
# TYPE parca_agent_native_unwinder_success_total counter
parca_agent_native_unwinder_success_total{unwinder="dwarf"} 23
parca_agent_native_unwinder_success_total{unwinder="dwarf_reach_bottom"} 40
parca_agent_native_unwinder_success_total{unwinder="dwarf_to_jit"} 0
parca_agent_native_unwinder_success_total{unwinder="jit_frame"} 0
parca_agent_native_unwinder_success_total{unwinder="jit_reach_bottom"} 0
parca_agent_native_unwinder_success_total{unwinder="jit_to_dwarf"} 0


# HELP parca_agent_profiler_bpf_maps_clean_errors_total Number of errors cleaning BPF maps
# TYPE parca_agent_profiler_bpf_maps_clean_errors_total counter
parca_agent_profiler_bpf_maps_clean_errors_total{map="dwarf_stack_traces",type="cpu"} 0
parca_agent_profiler_bpf_maps_clean_errors_total{map="process_info",type="cpu"} 0
parca_agent_profiler_bpf_maps_clean_errors_total{map="stack_counts",type="cpu"} 0
parca_agent_profiler_bpf_maps_clean_errors_total{map="stack_traces",type="cpu"} 0
parca_agent_profiler_bpf_maps_clean_errors_total{map="unwind_info_chunks",type="cpu"} 0

# HELP parca_agent_profiler_bpf_maps_refresh_proc_info_errors_total Number of errors refreshing process info
# TYPE parca_agent_profiler_bpf_maps_refresh_proc_info_errors_total counter
parca_agent_profiler_bpf_maps_refresh_proc_info_errors_total{error="hash",type="cpu"} 0
parca_agent_profiler_bpf_maps_refresh_proc_info_errors_total{error="unwind_table_add",type="cpu"} 0

# HELP parca_agent_profiler_events_lost_total Total number of profile events lost.
# TYPE parca_agent_profiler_events_lost_total counter
parca_agent_profiler_events_lost_total{type="cpu"} 0

# HELP parca_agent_profiler_events_received_total Total number of profile events received.
# TYPE parca_agent_profiler_events_received_total counter
parca_agent_profiler_events_received_total{event="empty",type="cpu"} 0
parca_agent_profiler_events_received_total{event="process_mappings",type="cpu"} 2887
parca_agent_profiler_events_received_total{event="refresh_proc_info",type="cpu"} 16
parca_agent_profiler_events_received_total{event="unwind_info",type="cpu"} 829

# HELP parca_agent_profiler_frame_drop_total Number of addresses dropped from the profile.
# TYPE parca_agent_profiler_frame_drop_total counter
parca_agent_profiler_frame_drop_total{reason="mapping_nil",type="cpu"} 28

# HELP parca_agent_profiler_map_read_attempts_total Number of attempts to read from the BPF maps.
# TYPE parca_agent_profiler_map_read_attempts_total counter
parca_agent_profiler_map_read_attempts_total{action="dwarf_unwind",stack="user",status="error",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="dwarf_unwind",stack="user",status="failed",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="dwarf_unwind",stack="user",status="missing",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="dwarf_unwind",stack="user",status="success",type="cpu"} 23
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="kernel",status="error",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="kernel",status="failed",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="kernel",status="missing",type="cpu"} 2580
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="kernel",status="success",type="cpu"} 290
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="user",status="error",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="user",status="failed",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="user",status="missing",type="cpu"} 0
parca_agent_profiler_map_read_attempts_total{action="kernel_unwind",stack="user",status="success",type="cpu"} 2847

# HELP parca_agent_profiler_profiles_drop_total Number of profiles dropped from the profile (one profile represents 1 process in a profiling duration).
# TYPE parca_agent_profiler_profiles_drop_total counter
parca_agent_profiler_profiles_drop_total{reason="process_info",type="cpu"} 24

# HELP parca_agent_profiler_stack_drop_total Total number of stacks dropped from the profile.
# TYPE parca_agent_profiler_stack_drop_total counter
parca_agent_profiler_stack_drop_total{reason="iterator",type="cpu"} 0
parca_agent_profiler_stack_drop_total{reason="read_kernel_stack",type="cpu"} 2580
parca_agent_profiler_stack_drop_total{reason="read_stack_count",type="cpu"} 0
parca_agent_profiler_stack_drop_total{reason="read_stack_count_zero",type="cpu"} 0
parca_agent_profiler_stack_drop_total{reason="read_stack_key",type="cpu"} 0
parca_agent_profiler_stack_drop_total{reason="read_user_stack_with_dwarf",type="cpu"} 0
parca_agent_profiler_stack_drop_total{reason="read_user_stack_with_frame_pointer",type="cpu"} 0

# HELP parca_agent_profiler_unwind_table_add_errors_total Total number of errors adding entries to the unwind table.
# TYPE parca_agent_profiler_unwind_table_add_errors_total counter
parca_agent_profiler_unwind_table_add_errors_total{error="need_more_rounds",type="cpu"} 0
parca_agent_profiler_unwind_table_add_errors_total{error="other",type="cpu"} 0
parca_agent_profiler_unwind_table_add_errors_total{error="procfs_race",type="cpu"} 446
parca_agent_profiler_unwind_table_add_errors_total{error="too_many_mappings",type="cpu"} 0

# HELP parca_agent_profiler_unwind_table_persist_errors_total Total number of errors persisting the unwind table.
# TYPE parca_agent_profiler_unwind_table_persist_errors_total counter
parca_agent_profiler_unwind_table_persist_errors_total{error="need_more_rounds",type="cpu"} 0
parca_agent_profiler_unwind_table_persist_errors_total{error="other",type="cpu"} 0

```

</p>
</details>

### CODE REVIEW

File: .github/workflows/container.yml [hunk 1]
```diff
@@ -148,9 +148,9 @@ jobs:
         shell: bash
         run: |
           ls -lR bpf
-          mkdir -p pkg/profiler/cpu/bpf
-          cp -r bpf/out/* pkg/profiler/cpu/bpf
-          rm pkg/profiler/cpu/bpf/*/pid_namespace.bpf.o
+          mkdir -p pkg/profiler/cpu/bpf/programs
```
Comment (javierhonduco) [inline .github/workflows/container.yml:151]: If we are doing this now, let's wipe the `cpu` part of the path as it's wrong. These are unwinders, the profiler type is directed by where are we attaching these programs

**Review by javierhonduco — APPROVED** [review]
Looks good! Could you provide the metrics in the test plan? Something like curling the metrics endpoint and showing that the counters we care about are there

File: .github/workflows/container.yml (hunk 1 above)
Comment (kakkoyun) [reply .github/workflows/container.yml:151]: So you want me to rename `pkg/profiler/cpu` to `pkg/profiler/unwinder` ?

**Comment by kakkoyun** [comment]
> Looks good! Could you provide the metrics in the test plan? Something like curling the metrics endpoint and showing that the counters we care about are there

Added them to description

File: .github/workflows/container.yml (hunk 1 above)
Comment (javierhonduco) [reply .github/workflows/container.yml]: That name works. This is somethign I wanted to do for the bpf progrs too in `bpf/` but let's try to focus this week on reliability 😄

Comment (kakkoyun) [reply .github/workflows/container.yml]: Let's talk about this later. We can change it to whatever we think is meaningful.

**Pull request merged**